### PR TITLE
fix(ui): Shop Panel Improvements

### PR DESCRIPTION
**Pull request by Koranir**
There was a bit of discussion on discord about how some people did not like smooth scrolling, so I've added an option to disable it.

The shop panel now also remembers which column was last selected, and tries to keep on it unless manually changed:

[remxp.webm](https://github.com/endless-sky/endless-sky/assets/101683475/1cffcd40-3bdb-4102-9bca-95e29be326e2)

### CODE REVIEW

File: source/ShopPanel.cpp [hunk 1]
```diff
@@ -64,18 +64,19 @@ namespace {
 	}
 
 	// Update smooth scroll towards scroll.
-	void UpdateSmoothScroll(const double scroll, double &smoothScroll)
+	double UpdateSmoothScroll(const double scroll, const double smoothscroll)
```
Comment (TomGoodIdea) [inline source/ShopPanel.cpp:67]: please keep the camel

**Comment by ziproot** [comment]
See https://github.com/endless-sky/endless-sky/pull/9504#issuecomment-1776383384

**Comment by quyykk** [comment]
> I've added an option to disable it.

I don't think that's a good idea, we shouldn't be adding preferences for every small thing.

**Comment by Koranir** [comment]
> > I've added an option to disable it.
> 
> I don't think that's a good idea, we shouldn't be adding preferences for every small thing.

There have been a lot of requests to allow disabling the smooth scrolling.
Mobile doesn't need it, and it just ends up just adding delay to scrolls, and some people don't like how it looks.

**Comment by quyykk** [comment]
> There have been a lot of requests to allow disabling the smooth scrolling.
>
> Mobile doesn't need it, and it just ends up just adding delay to scrolls, and some people don't like how it looks.

In that case, let's improve the scrolling by either fixing the smoothness thing people don't like or just simply removing it.

**Comment by Koranir** [comment]
> > There have been a lot of requests to allow disabling the smooth scrolling.
> > Mobile doesn't need it, and it just ends up just adding delay to scrolls, and some people don't like how it looks.
> 
> In that case, let's improve the scrolling by either fixing the smoothness thing people don't like or just simply removing it.

I don't understand what you mean, some people don't like their scrolls being smoothed, some people do, why remove it? And the "smoothness thing people don't like" is the whole concept of smooth scrolling.

I myself prefer not having the screen jump around everywhere, but some people like it being more responsive and mobile doesn't need it. It's subjective, which is the exact use-case of which a preference is for.

File: data/_ui/tooltips.txt [hunk 2]
```diff
@@ -1350,5 +1350,8 @@ tip "Landing zoom"
 tip "Scroll speed"
 	`The speed of scrolling in panels that can scroll.`
 
+tip "Scroll smoothing"
+	`Scrolling is smoothed in the Outfitter & Shipyard`
```
Comment (Amazinite) [inline data/_ui/tooltips.txt:1354]:
```suggestion
	`Whether scrolling is smoothed in the outfitter and shipyard.`
```

**Comment by Amazinite** [comment]
In reviewing different locations in the game that can be scrolled for https://github.com/endless-sky/endless-sky/pull/9603#issuecomment-1859216316 while taking a close look at our UI consistency, I noticed how if you jump around between the different places that can be scrolled, the shop currently sticks out like a sore thumb being the only place with smooth scrolling. For the sake of consistency, I think we should go all or nothing on whether we use smooth scrolling, and in that sense it's easier to just toss smooth scrolling from the shop than it is to add smooth scrolling everywhere else. So at the very least I think that if we were to add a preference for smooth scrolling, it should be default off. But, @quyykk, @Saugia, and (I believe) @warp-core would rather just see smooth scrolling removed entirely than made a setting.

While it might be that some people like the smooth scrolling over the standard scrolling, it isn't exactly some flagship feature that people are coming to the game for, so I think it's fine to toss it entirely rather than maintain both standard and smooth scrolling.

**Comment by thewierdnut** [comment]
>> There have been a lot of requests to allow disabling the smooth scrolling.
>
> Mobile doesn't need it as it just ends up just adding delay to scrolls for them, and some people don't like how it looks.

The reason I added smooth scrolling to the mobile version was because when you used a gamepad, (or keyboard controls for that matter) the cursor would skip around and it was difficult to tell what was going on. I went through several iterations of the scrolling before I found something that I thought worked right. Some observations that I made:

1. The smoothing function doesn't add much. A linear interpolation was just as effective at indicating what direction the cursor just jumped. 
2. The number of animated frames needs to be kept small, somewhere on the order of 5 frames, maybe 10 to 15 for large jumps. The purpose isn't to look cool, but to indicate to the user which direction the cursor just jumped.
3. Smooth scrolling should be turned off for drag operations. Its annoying to have the the screen lag behind the mouse (or finger) when dragging the pane up and down.

**Comment by Amazinite** [comment]
> Smooth scrolling should be turned off for drag operations. Its annoying to have the the screen lag behind the mouse (or finger) when dragging the pane up and down.

I think this is where the main contention with smooth scrolling in the shop is coming from, because I believe it's still active when dragging with the mouse.

**Comment by Koranir** [comment]
Gonna close this for now, put the non-contended bit into a separate PR. I personally think smooth scrolling should be in all the scrollable panels (and optional), but I'll probably wait for #9603 and its `ScrollVar` for that.

**Pull request closed without merging**